### PR TITLE
[Validator] added BIC validator

### PR DESCRIPTION
**Pull request by mvhirsch**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes, PR: https://github.com/symfony/symfony/pull/15519 |
| Applies to | 2.8 |
| Fixed tickets | none |

### CODE REVIEW

File: reference/constraints/Bic.rst [hunk 1]
```diff
@@ -0,0 +1,107 @@
+Bic
+===
+
+.. versionadded:: 2.8
+    The Bic constraint was introduced in Symfony 2.8.
+
+This constraint is used to ensure that a value has the proper format of an
+`Business Identifier Code (BIC)`. BIC is an internationally agreed means to
+uniquely identify both financial and non-financiel institutions.
```
Comment (javiereguiluz) [inline reference/constraints/Bic.rst]: `non-financiel` -> `non-financial`

Comment (mvhirsch) [inline reference/constraints/Bic.rst]: you're absolutely right!

File: reference/constraints/Bic.rst [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+| Class          | :class:`Symfony\\Component\\Validator\\Constraints\\Bic`              |
++----------------+-----------------------------------------------------------------------+
+| Validator      | :class:`Symfony\\Component\\Validator\\Constraints\\BicValidator`     |
++----------------+-----------------------------------------------------------------------+
+
+Basic Usage
+-----------
+
+To use the Bic validator, simply apply it to a property on an object that
+will contain a Business Identifier Code.
```
Comment (OskarStark) [inline reference/constraints/Bic.rst]: ```will contain a Business Identifier Code (BIC).```

**Comment by OskarStark** [comment]
looks good to me :+1: 

thank you @mvhirsch

File: reference/constraints/Bic.rst [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+-----------------
+
+message
+~~~~~~~
+
+**type**: ``string`` **default**: ``This is not a valid Business Identifier Code (BIC).``
+
+The default message supplied when the value does not pass the Bic check.
+
+.. include:: /reference/constraints/_payload-option.rst.inc
```
Comment (OskarStark) [inline reference/constraints/Bic.rst]: i'm not sure why, but this part is not in the 2.8 branch:
https://github.com/symfony/symfony-docs/blob/2.8/reference/constraints/Iban.rst

any idea @javiereguiluz... confused... :smile:

File: reference/constraints.rst [hunk 4]
```diff
@@ -52,6 +52,7 @@ Validation Constraints Reference
    constraints/Iban
    constraints/Isbn
    constraints/Issn
+   constraints/Bic
```
Comment (OskarStark) [inline reference/constraints.rst]: in case this is not an alphanumeric order, i would put this directly under Iban, because they belong togehter

File: reference/constraints/Bic.rst (hunk 3 above)
Comment (javiereguiluz) [inline reference/constraints/Bic.rst]: "Upwards branch merging" is done from time to time in a non-formal schedule. Sometimes 2.8 and master are behind the other branchs, but at the end our doc managers always merge upwards.

Comment (mvhirsch) [inline reference/constraints/Bic.rst]: Sure it is ;-)

https://raw.githubusercontent.com/symfony/symfony-docs/2.8/reference/constraints/Iban.rst

Comment (OskarStark) [inline reference/constraints/Bic.rst]: thank you guys :+1:

**Comment by mvhirsch** [comment]
@OskarStark I moved the BIC near the IBAN validation.

**Comment by OskarStark** [comment]
thank you @mvhirsch

File: reference/constraints/Bic.rst [hunk 5]
```diff
@@ -0,0 +1,107 @@
+Bic
+===
+
+.. versionadded:: 2.8
+    The Bic constraint was introduced in Symfony 2.8.
+
+This constraint is used to ensure that a value has the proper format of an
+`Business Identifier Code (BIC)`. BIC is an internationally agreed means to
```
Comment (WouterJ) [inline reference/constraints/Bic.rst]: the closing `` ` `` needs to be followed by an underscore: `` `Business Identifier Code (BIC)`_``

File: reference/constraints/Bic.rst [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+-----------
+
+To use the Bic validator, simply apply it to a property on an object that
+will contain a Business Identifier Code.
+
+.. configuration-block::
+
+    .. code-block:: php-annotations
+
+        // src/Acme/SubscriptionBundle/Entity/Transaction.php
```
Comment (WouterJ) [inline reference/constraints/Bic.rst]: please use `src/AppBundle/Entity/Transaction.php` (this probably is inconsistent with the other articles, but they need to be updated)

File: reference/constraints/Bic.rst [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        namespace Acme\SubscriptionBundle\Entity;
```
Comment (WouterJ) [inline reference/constraints/Bic.rst]: that would of course mean that you have to change this as well (as well as the XML, Yaml and PHP code examples)

File: reference/constraints/Bic.rst (hunk 3 above)
Comment (WouterJ) [inline reference/constraints/Bic.rst]: @OskarStark don't rely on GitHub's reStructuredText rendering. It doesn't support lots of the things that we use.

File: reference/constraints/Bic.rst [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+Available Options
+-----------------
+
+message
+~~~~~~~
+
+**type**: ``string`` **default**: ``This is not a valid Business Identifier Code (BIC).``
+
+The default message supplied when the value does not pass the Bic check.
```
Comment (WouterJ) [inline reference/constraints/Bic.rst]: "[...] BIC check" ?

Comment (OskarStark) [inline reference/constraints/Bic.rst]: :+1:

**Comment by mvhirsch** [comment]
@WouterJ I just updated the docs according to your notes. Thank you very much!

File: reference/constraints.rst [hunk 9]
```diff
@@ -50,6 +50,7 @@ Validation Constraints Reference
    constraints/Currency
    constraints/Luhn
    constraints/Iban
+   constraints/Bic
```
Comment (xabbuh) [inline reference/constraints.rst]: This should also be added to `reference/constraints.map.rst.inc`.

File: reference/constraints/Bic.rst [hunk 10]
```diff
@@ -0,0 +1,107 @@
+Bic
+===
+
+.. versionadded:: 2.8
+    The Bic constraint was introduced in Symfony 2.8.
+
+This constraint is used to ensure that a value has the proper format of an
```
Comment (xabbuh) [inline reference/constraints/Bic.rst]: [...] of a [...]

File: reference/constraints/Bic.rst [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+`Business Identifier Code (BIC)`_. BIC is an internationally agreed means to
+uniquely identify both financial and non-financial institutions.
+
++----------------+-----------------------------------------------------------------------+
+| Applies to     | :ref:`property or method<validation-property-target>`                 |
```
Comment (xabbuh) [inline reference/constraints/Bic.rst]: There should be a space before the opening angle bracket.

File: reference/constraints/Bic.rst [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+-----------
+
+To use the Bic validator, simply apply it to a property on an object that
+will contain a Business Identifier Code (BIC).
+
+.. configuration-block::
+
+    .. code-block:: php-annotations
+
+        // src/Acme/AppBundle/Entity/Transaction.php
```
Comment (xabbuh) [inline reference/constraints/Bic.rst]: The `Acme` part of the namespace and directory should be completely removed.

File: reference/constraints/Bic.rst [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        namespace Acme\AppBundle\Entity;
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Transaction
+        {
+            /**
+             * @Assert\Bic(
+             *     message="This is not a valid Business Identifier Code (BIC)."
```
Comment (xabbuh) [inline reference/constraints/Bic.rst]: I think we can omit the message (it's already the default message by the way).

File: reference/constraints.rst (hunk 9 above)
Comment (mvhirsch) [inline reference/constraints.rst]: I'm sorry, but I can't find such file. Did you mean `reference/constraints/map.rst.inc` instead?

File: reference/constraints/Bic.rst (hunk 12 above)
Comment (mvhirsch) [inline reference/constraints/Bic.rst]: Will the `Acme` part be removed in all other `reference/constraints/*.rst` files in 2.8?
Otherwise the `Bic.rst` would be inconsistent to the rest of the documentation.

File: reference/constraints.rst (hunk 9 above)
Comment (xabbuh) [inline reference/constraints.rst]: Sorry for the confusion. That's indeed the file I meant.

File: reference/constraints/Bic.rst (hunk 12 above)
Comment (xabbuh) [inline reference/constraints/Bic.rst]: I think we should update the other parts too.

File: reference/constraints.rst (hunk 9 above)
Comment (WouterJ) [inline reference/constraints.rst]: But let's do that in another PR, which is based on 2.3. For now, only update the newly added code examples

Comment (xabbuh) [inline reference/constraints.rst]: I think you mean updating the code examples, right? And yes, I agree with you then.

**Comment by mvhirsch** [comment]
I removed the *Acme* namespace and added *Bic* to the map.rst.inc.

File: reference/constraints/map.rst.inc [hunk 14]
```diff
@@ -66,6 +66,7 @@ File Constraints
 Financial and other Number Constraints
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+* :doc:`Bic` </reference/constraints/Bic>`
```
Comment (xabbuh) [inline reference/constraints/map.rst.inc]: The backtick in the middle needs to be remove. Otherwise, the path after it would not be treated as part of the `doc` directive.

Comment (mvhirsch) [inline reference/constraints/map.rst.inc]: I'm sorry, was a typo, just fixed that. Thank you!

**Comment by OskarStark** [comment]
i think we can add the ```Finished``` tag ;-)

**Comment by WouterJ** [comment]
@OskarStark as the code PR isn't merged yet, we'll keep the "On hold" label (meaning: waiting code merge).

**Comment by xabbuh** [comment]
The code has been merged. Congrats @mvhirsch!

**Comment by WouterJ** [comment]
:+1:

**Comment by xabbuh** [comment]
Oh, I forgot to :+1:.

**Comment by xabbuh** [comment]
Thank you Michael. I have merged your pull request and made a tiny tweak in f64d406121e6358e89f70db65472538dad594e0c (removing the error message which was the default anyway).
